### PR TITLE
Correct exponent_max to conform to IEEE754-2008

### PR DESCRIPTION
**Pull request by mschauer**
`exponent_max` was not giving the maximum exponent, but a sentinel value for `Inf` and `NaN` which is not a legal exponent. I found this error prone. Exponent exponent_max returns actual maximum exponent. As a by-product the code become somewhat more readable. ("if k is one bigger than the maximum exponent, first multiply y with 2 and then with the highest power of two representable...") I added internal docstrings to help the future readers.

### CODE REVIEW

**Comment by simonbyrne** [comment]
Do we need this? What's wrong with `exponent(realmax(T))`?

**Comment by simonbyrne** [comment]
See also #24631

**Comment by simonbyrne** [comment]
I would be in favour of renaming `exponent_max` to `exponent_raw_max` though.

**Comment by JeffreySarnoff** [comment]
Julia has been assigning exponent_max values that do not conform to IEEE754-2008,

`exponent_max` (*Emax* in the Standard) has a tabulated value for each floating point type.

```julia
#   These are the tabulated values (see Table 3.2 on page 8 of the IEEE Standard).

exponent_max(::Type{Float16}) =   15
exponent_max(::Type{Float32}) =  127
exponent_max(::Type{Float64}) = 1023
```
All the details are here: https://github.com/JeffreySarnoff/IEEE754-2008.md.

**Comment by mschauer** [comment]
`exponent(realmax(T))` does give *Emax*, but there is some cognitive ballast asserting that it does to it.

**Comment by JeffreySarnoff** [comment]
As I understand the writing, *Emax* is one of their essential quantities.  Since *Emin* is, by design,  1-*Emax* always, and that almost evenly biased exponents are used,  *Emax* is another way of writing exponent_bias (*bias*).  The fact that `exponent(realmax(T))` also resolves as *Emax(T)* is more a consistency check than definitional -- I agree with your observation.

**Comment by mschauer** [comment]
Ping

**Comment by JeffreySarnoff** [comment]
the freebsd link is to nowhere

**Comment by mschauer** [comment]
A helpful fairy fixed the tests.

**Comment by JeffreySarnoff** [comment]
super -- I await the merge!

**Comment by mschauer** [comment]
This is still good to go.

**Comment by JeffreySarnoff** [comment]
^ (any reason not to merge this)?

**Comment by simonbyrne** [comment]
I'm in favour of this. Since it is a change, should probably be documented in NEWS.md

**Comment by mschauer** [comment]
Yes. BTW pkg.julialang.com shows no use, only @JeffreySarnoff providing the functionality independently in https://github.com/JeffreySarnoff/LowLevelFloatFunctions.jl

**Comment by mschauer** [comment]
What is the right header for the NEW item? 

*  The unexported and previously undocumented function `exponent_max` now returns the actual maximum exponent according to IEEE754-2008, not the sentinel value.

**Comment by simonbyrne** [comment]
Ah, if it's not exported it probably doesn't need an item.

File: base/math.jl [hunk 1]
```diff
@@ -37,13 +37,28 @@ end
                                 "Complex(x)^y, or similar.")))
 end
 
+"""
+    exponent_max(T)
+
+Maximum exponent a real floating point number of type `T` can have,
```
Comment (simonbyrne) [inline base/math.jl]:
```suggestion
Maximum value of [`exponent`](@ref) for a real floating point number of type `T`.
```

**Review by simonbyrne — COMMENTED** [review]
Would be good to link to `exponent` help.

**Comment by mschauer** [comment]
Something with the CI is failing currently, I think unrelatedly to this change. I'll leave it and maybe someone with rights can trigger the tests when things run smoothly again.

**Comment by mschauer** [comment]
Ping

**Comment by musm** [comment]
Needs a rebase

**Comment by mschauer** [comment]
I am out, honestly

**Comment by JeffreySarnoff** [comment]
@admin fyi

**Comment by musm** [comment]
@mschauer Sorry I didn't understand this comment? Do you mean you are no longer interested in rebasing this? Thanks.

**Comment by StefanKarpinski** [comment]
That was my interpretation of the comment and frankly it's pretty understandable given that it's been three years.

**Comment by JeffreySarnoff** [comment]
I concur. Would an appropriate admin rebase this, please.

**Comment by musm** [comment]
I'll try to get around to rebasing this soon. Thanks for the effort @mschauer

**Comment by musm** [comment]
https://github.com/JuliaLang/julia/pull/38929/files

**Pull request closed without merging**